lib/src/component/backend.rs

```diff
@@ -0,0 +1,139 @@
+use {
+    super::fastly::api::{backend, http_types},
+    super::FastlyError,
+    crate::{error::Error, session::Session},
+};
+
+#[async_trait::async_trait]
+impl backend::Host for Session {
+    async fn exists(&mut self, backend: String) -> Result<bool, FastlyError> {
+        Ok(self.backend(&backend).is_some())
+    }
+
+    async fn is_healthy(&mut self, backend: String) -> Result<backend::BackendHealth, FastlyError> {
+        // just doing this to get a different error if the backend doesn't exist
+        let _ = self.backend(&backend).ok_or(Error::InvalidArgument)?;
+        Ok(backend::BackendHealth::Unknown)
+    }
+
+    async fn is_dynamic(&mut self, backend: String) -> Result<bool, FastlyError> {
+        if self.dynamic_backend(&backend).is_some() {
+            Ok(true)
+        } else if self.backend(&backend).is_some() {
+            Ok(false)
+        } else {
+            Err(Error::InvalidArgument.into())
+        }
+    }
+
+    async fn get_host(&mut self, backend: String, max_len: u64) -> Result<String, FastlyError> {
+        let backend = self.backend(&backend).ok_or(Error::InvalidArgument)?;
+
+        let host = backend.uri.host().expect("backend uri has host");
+
+        if host.len() > usize::try_from(max_len).unwrap() {
+            return Err(Error::BufferLengthError {
+                buf: "host_out",
+                len: "host_max_len",
+            }
+            .into());
+        }
+
+        Ok(String::from(host))
+    }
+
+    async fn get_override_host(
+        &mut self,
+        backend: String,
+        max_len: u64,
+    ) -> Result<Option<String>, FastlyError> {
+        let backend = self.backend(&backend).ok_or(Error::InvalidArgument)?;
+        if let Some(host) = backend.override_host.as_ref() {
+            let host = host.to_str()?;
+
+            if host.len() > usize::try_from(max_len).unwrap() {
+                return Err(Error::BufferLengthError {
+                    buf: "host_out",
+                    len: "host_max_len",
+                }
+                .into());
+            }
+
+            Ok(Some(String::from(host)))
+        } else {
+            Ok(None)
```

Comment:
We could change this to returning `Err(Error::ValueAbsent)`, but it felt a little bit nicer to encode that case as `Ok(None)` instead. Happy to return to the old behavior though, as it would pull some logic out of the adapter for translating the result.